### DISCUSSION TITLE
Timer with socket

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#10793] [General]
I have a problem with this little program.
```
import socket,network
from machine import Timer
import sys
from time import sleep

def config_wifi():
    sta=network.WLAN(network.STA_IF)
    sta.active(True)
    sta.ifconfig(('192.168.144.3','255.255.255.0','255.255.255.0','8.8.8.8'))
    sta.connect('MicroPython-AP','123456789')
    max_cont=5
    cont=0
    while sta.isconnected()==False and cont < max_cont:
        print(cont)
        cont +=1
        sleep(1)
        if cont == max_cont:
            sys.exit()
    return

def exit_door():
    pass #do something
    
config_wifi()
tim = Timer(-1)
adress_server=socket.getaddrinfo("192.168.144.4",800)[0][-1]
com_net=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
com_net.connect(adress_server)
com_net.sendall(b'test')    # Send This message to server
tim.init(period=5000, mode=Timer.ONE_SHOT, callback=exit_door)
data = com_net.recv(1024)	    # Now, receive the echoed data from server
tim.deinit()
com_net.close() # close the connection
```
It's an ESP8266 client that connects to a server over wifi.
But if the server has not configured its socket, the client remains stuck on com_net.recv(1024). (line 32)
I was thinking of putting a timer that would unlock after 5 seconds, but I got this error TypeError: function takes 0 positional arguments but 1 were given.
When I do Ctrl-C after 10 seconds, I have File stdin, line 32, in module
KeyboardInterrupt:
Which shows that com_net.recv is blocking. Since there is no error message, I cannot circumvent the problem with a Try/Except.
Do you have a solution for me?
Thanks in advance

**Comment by robert-hh** [comment]
The timer callback function requires an argument, which is the timer object. You can ignore it in the function, but itusz ne present.
```
def exit_door(t):
    pass
```

**Reply by PatLub85** [reply]
Done but it doesn't change anything

**Reply by robert-hh** [reply]
But the error `TypeError: function takes 0 positional arguments but 1 were given.` must be gone if the function `exit_door() `has a formal argument.

**Comment by bixb922** [comment]
One option is to use socket.setblocking to set the socket for non-blocking operation. Then loop doing recv, sleeping and testing time elapsed until you have received all 1024 bytes sent by your server or the maximum time has exceeded.

Or take a look at socket.settimeout if supported on your port (https://docs.micropython.org/en/latest/library/socket.html).

**Reply by PatLub85** [reply]
Now with setblocking(0) I get error [Errno 11] EAGAIN.
If I do a try/except on this error, I can get out of the block and I no longer need the timer.
It's not what I expected, but it meets my need.

**Reply by PatLub85** [reply]
I modified my program according to your remarks and here are the different cases observed.
```
import socket,network
from machine import Timer
import sys
from time import sleep

def config_wifi():
    sta=network.WLAN(network.STA_IF)
    sta.active(True)
    sta.ifconfig(('192.168.144.3','255.255.255.0','255.255.255.0','8.8.8.8'))
    sta.connect('MicroPython-AP','123456789')
    max_cont=5
    cont=0
    while sta.isconnected()==False and cont < max_cont:
        print(cont)
        cont +=1
        sleep(1)
        if cont == max_cont:
            sys.exit()
    return

def exit_door(t):
    print('exit_door')
    sys.exit()
    
config_wifi()
adress_server=socket.getaddrinfo("192.168.144.4",800)[0][-1]
com_net=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
com_net.connect(adress_server)
tim = Timer(-1)
com_net.setblocking(0)
com_net.sendall('START')    # Send This message to server
tim.init(period=5000, mode=Timer.ONE_SHOT, callback=exit_door)
try:
    data = com_net.recv(1024)	    # Now, receive the echoed data from server
except OSError as e:
    print('error',e)
    exit_door(e)
tim.deinit()
com_net.close() # close the connection
```
- case 1: the client cannot connect to the server's wifi, the error is handled by try/except
- case 2: the server socket responds, the client sends its 'START' message, the timer starts, no response from the server, and error management by try/except. The program ends with a sys.exit and then the timer tries to run and gets stuck
```
The messages :
error [Errno 11] EAGAIN
exit_door
>>> exit_door
Traceback (most recent call last):
  File "<stdin>", line 23, in exit_door
SystemExit: 
 ```
I'm lost

**Reply by bixb922** [reply]
Did you try making a loop as suggested? 
Look up EAGAIN, it means "there is no data available right now, try again later".
Use time.ticks_ms() and time.ticks_diff() to measure elapsed time ()https://docs.micropython.org/en/latest/library/time.html)

**Reply by PatLub85** [reply]
I found a solution that works for my use
```
        com_net.sendall('START')   
        try:
            com_net.settimeout(1.0)  # The solution !
            data = com_net.recv(1024)	   
        except OSError as e:
            print('error',e)
            com_net.close()
            exit_door()
        com_net.close() 
```
Thanks for your help
